### PR TITLE
Allow multiple cudagraph recordings per compiled graph

### PR DESCRIPTION
**Pull request by mlazos**
### Introduction/Problem

Today when dynamo traces a builtin nn module (nn.Linear for example) it will specially handle parameters of that module by storing them as constant attributes of the graph. This requires that dynamo guard on the ID of the NNModule because if the instance of the module changes, we need to retrace and recollect the new parameters as attributes of the graph. This creates a 1:1 compiled graph to cudagraph relationship.

With hierarchical compilation, dynamo will treat builtin nn modules like any other code. This reduces complexity and critically, if there are multiple identical layers in a model, we only need to compile one of those layers once, and reuse the same compiled artifact for each layer. This introduces a problem for the current approach to parameter handling. Since the parameters could now possibly change across calls to the compiled artifact, these need to be inputs to the graph instead of attributes. This introduces a problem for cudagraphs - previously cudagraphs was guaranteed that the parameters of builtin NN Modules would be constant across calls, but now since the compiled artifact needs to be agnostic to the actual instance of the NN module being used these parameter memory locations may vary. Previously cudagraphs simply copies varying inputs to cudagraph owned memory, but since the parameters are quite large, this is catastrophic for performance. 

### Solution
To avoid this performance cliff, this PR allows cudagraphs to re-record a new cudagraph if only parameters change. Metadata about which arguments are parameters are propagated from AOT Autograd to compile_fx, and these indices are passed to cudagraphs. If these memory locations change, a new graph is recorded vs previously where this would be an error (because this previously should not happen). This enables a 1:many compiled graph to cudagraph relationship. Across similar modules we will re-record cudagraphs and dispatch the correct graph if parameter pointers match when the cudagraph is executed.

### Next steps (if needed)
It is theoretically possible that a user passes Parameters that change frequently as inputs to model code - if this is a common issue this design allows for dynamo to pass metadata indicating which parameters were created in a builtin NN Module context to only permit those parameters to have the multi-cudagraph behavior, but this PR does not implement this. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126822
* #126821
* #126820

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
would you mind updating description with the problem and solution?

**Review by eellison — COMMENTED** [review]
Looks good ! just one comment, would like to only re-record parameters in hierarchical compilatino.

> It is theoretically possible that a user passes Parameters that change frequently as inputs to model code - if this is a common issue this design allows for dynamo to pass metadata indicating which parameters were created in a builtin NN Module context to only permit those parameters to have the multi-cudagraph behavior, but this PR does not implement this.

This is an issue we have run into before. Can we specify when a model is compiled with hierarchical compilation and only turn on the multi-graph behavior for parameters ?

**Comment by mlazos** [comment]
> Looks good ! just one comment, would like to only re-record parameters in hierarchical compilatino.
> 
> > It is theoretically possible that a user passes Parameters that change frequently as inputs to model code - if this is a common issue this design allows for dynamo to pass metadata indicating which parameters were created in a builtin NN Module context to only permit those parameters to have the multi-cudagraph behavior, but this PR does not implement this.
> 
> This is an issue we have run into before. Can we specify when a model is compiled with hierarchical compilation and only turn on the multi-graph behavior for parameters ?

So today if inline_builtin_nn_modules is False parameters for builtin nn modules will still be added as constants on the graph and bypass this (since they will no longer be inputs). However if a user manually passed a parameter as an input, we would re-record if this changes regardless of the value of inline_builtin_nn_modules. I can check this config value and only add indices if the flag is True. Eventually True will be the default value FYI though.

**Review by eellison — COMMENTED** [review]
Just to clarify one thing - they won't be constants, they are lifted attributes from the dynamo graph.

Yea, I would to make this change because otherwise if you have parameter which is changing addresses we would never warn or error. It's better to error with data ptr changed than continually recapture.

File: torch/_inductor/cudagraph_trees.py [hunk 1]
```diff
@@ -1558,6 +1558,12 @@ def check_invariants(self, inputs: List[Tensor]) -> bool:
         ):
             return False
 
+        # static input data pointers remain stable
+        if not torch._C._tensors_data_ptrs_at_indices_equal(
+            inputs, self.static_input_data_ptrs, self.static_input_idxs
+        ):
+            return False
```
[the remark this answers is not in the document]
Comment (mlazos) [reply torch/_inductor/cudagraph_trees.py]: I set it on self and reset if the node is run

**Comment by mlazos** [comment]
> Just to clarify one thing - they won't be constants, they are lifted attributes from the dynamo graph.
> 
> Yea, I would to make this change because otherwise if you have parameter which is changing addresses we would never warn or error. It's better to error with data ptr changed than continually recapture.

discussed offline - gated this change on the flag, once this flag is on by default, I will add the restriction in dynamo that we only treat builtin nn module params as static

File: torch/_inductor/cudagraph_trees.py [hunk 2]
```diff
@@ -753,6 +753,8 @@ def __init__(
         self.device = device_index
         self.stack_traces = stack_traces
         self.stream = stream
+        # Used to only check this once
+        self.static_inputs_stable = False
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py:757]: I don't think we need to update this from True/False. 

If CUDAGraphNode is instantiated with `torch._dynamo.config.inline_inbuilt_nn_modules = True`:

- in check invariants, check static data ptrs
- skip check_static_inputs_are_stable

If False:

- dont check static data ptrs in check invariants
- do check_static_inputs_are_stable

Also add comment about what it represents

File: torch/_inductor/compile_fx.py [hunk 3]
```diff
@@ -120,6 +120,18 @@ def complex_memory_overlap(t: torch.Tensor) -> bool:
...
 
 
+def get_static_input_idxs(num_fixed):
+    # If we are inlining NNModules, we treat all torch.nn.Parameters as static for the purposes
+    # of cudagraphs. Rather than copying these into cudagraph-owned memory
+    # like we do for normal inputs on each run, we will re-record a cudagraph if these
+    # parameter locations change.
+    context = torch._guards.TracingContext.try_get()
+    if not context or not context.fw_metadata:
+        return []
```
Comment (eellison) [inline torch/_inductor/compile_fx.py:130]: nit.. should we just return list(range(num_fixed)) here ?  dont know if that will ever happen

**Review by eellison — APPROVED** [review]
Looks good! just a few small comments. nice tests!

File: torch/_inductor/cudagraph_trees.py (hunk 2 above)
Comment (mlazos) [reply torch/_inductor/cudagraph_trees.py:757]: yeah got it, we can just burn this in

File: torch/_inductor/compile_fx.py [hunk 4]
```diff
@@ -118,6 +118,18 @@ def complex_memory_overlap(t: torch.Tensor) -> bool:
     return False
 
 
+def get_static_input_idxs(num_fixed):
```
[the remark this answers is not in the document]
Comment (xmfan) [reply torch/_inductor/compile_fx.py:123]: Land this first. Mine was just a BE

**Comment by mlazos** [comment]
pytorchbot merge -f "smoketest perf queue is a day behind, this is gated by a flag and all tests are passing"

**Pull request closed without merging**